### PR TITLE
Migrate alertmanager configuration

### PR DESCRIPTION
**Pull request by QuentinBisson**
Replaces https://github.com/giantswarm/prometheus-meta-operator/pull/1192/files towards https://github.com/giantswarm/giantswarm/issues/25137

This PR is moves the Alertmanager global configuration generated by PMO to an AlertmanagerConfig CR applied by the helm chart.

The commit history should help reviewing this PR

## Checklist

I have:

- [x] Described why this change is being introduced
- [x] Separated out refactoring/reformatting in a dedicated PR
- [x] Updated changelog in `CHANGELOG.md`

### CODE REVIEW

File: helm/prometheus-meta-operator/values.yaml [hunk 1]
```diff
@@ -26,6 +26,31 @@ alertmanager:
...
+  - name: atlas
+    oncall: true
+    stableNotifyOnly: true
+    extraMatchers:
+    - name: type
+      matchType: "!="
+      value: "heartbeat"
+    - name: alertname
+      matchType: "!~"
+      value: "Inhibition.*"
```
Comment (TheoBrigitte) [inline helm/prometheus-meta-operator/values.yaml:39]: Feels weird that Atlas only has those extra matchers and not other teams, I guess no one want to receive heartbeat or inhibition.

**Review by TheoBrigitte — COMMENTED** [review]
It's a lot of changes even going through commit history. I guess files moves around and changed a bit, but I cannot tell what changed.

File: helm/prometheus-meta-operator/values.yaml (hunk 1 above)
Comment (QuentinBisson) [reply helm/prometheus-meta-operator/values.yaml:39]: exactly

File: flag/service/service.go [hunk 2]
```diff
@@ -3,27 +3,23 @@ package service
...
 	"github.com/giantswarm/prometheus-meta-operator/v2/flag/service/prometheus"
 	"github.com/giantswarm/prometheus-meta-operator/v2/flag/service/provider"
 	"github.com/giantswarm/prometheus-meta-operator/v2/flag/service/security"
-	"github.com/giantswarm/prometheus-meta-operator/v2/flag/service/slack"
 	"github.com/giantswarm/prometheus-meta-operator/v2/flag/service/vault"
 )
 
 // Service is an intermediate data structure for command line configuration flags.
 type Service struct {
-	Grafana      grafana.Grafana
```
Comment (QuentinBisson) [inline flag/service/service.go:19]: Deleting Grafana and Slack flags as they are no longer necessary

**Comment by QuentinBisson** [comment]
Diff on Gauss between old and new config:

```colordiff
colordiff old-alertmanager.yaml new-alertmanager.yaml
10a11
>   opsgenie_api_key: <secret>
16c17
<   receiver: root
---
>   receiver: monitoring/alertmanager-config/root
35,37c36,38
<   - receiver: harbor_implementation
<     matchers:
<     - app="harbor"
---
>   - receiver: monitoring/alertmanager-config/harbor_implementation
>     match:
>       app: harbor
39c40
<   - receiver: falco_noise_slack
---
>   - receiver: monitoring/alertmanager-config/falco_noise_slack
43c44,46
<   - receiver: team_shield_slack
---
>   - receiver: monitoring/alertmanager-config/team_shield_slack
>     match:
>       team: shield
46d48
<     - team="shield"
48,50c50,52
<   - receiver: opsgenie_router
<     matchers:
<     - severity="page"
---
>   - receiver: monitoring/alertmanager-config/opsgenie_router
>     match:
>       severity: page
52,54c54,56
<   - receiver: team_phoenix_slack
<     matchers:
<     - alertname="ServiceLevelBurnRateTooHigh"
---
>   - receiver: monitoring/alertmanager-config/team_phoenix_slack
>     match:
>       alertname: ServiceLevelBurnRateTooHigh
56c58,60
<   - receiver: team_atlas_slack
---
>   - receiver: monitoring/alertmanager-config/team_atlas_slack
>     match:
>       team: atlas
60d63
<     - team="atlas"
63,75c66,68
<   - receiver: team_phoenix_slack
<     matchers:
<     - severity=~"page|notify"
<     - sloth_severity=~"page|ticket"
<     - team="celestial"
<     continue: false
<   - receiver: team_phoenix_slack
<     matchers:
<     - severity=~"page|notify"
<     - sloth_severity=~"page|ticket"
<     - team="firecracker"
<     continue: false
<   - receiver: team_phoenix_slack
---
>   - receiver: monitoring/alertmanager-config/team_bigmac_slack
>     match:
>       team: bigmac
78,79d70
<     - sloth_severity=~"page|ticket"
<     - team="phoenix"
81c72,74
<   - receiver: team_hydra_slack
---
>   - receiver: monitoring/alertmanager-config/team_clippy_slack
>     match:
>       team: clippy
84d76
<     - team="hydra"
86c78,80
<   - receiver: team_bigmac_slack
---
>   - receiver: monitoring/alertmanager-config/team_hydra_slack
>     match:
>       team: hydra
89d82
<     - team="bigmac"
91c84,86
<   - receiver: team_clippy_slack
---
>   - receiver: monitoring/alertmanager-config/team_phoenix_slack
>     match:
>       team: phoenix
94d88
<     - team="clippy"
96c90,92
<   - receiver: team_rocket_slack
---
>   - receiver: monitoring/alertmanager-config/team_rocket_slack
>     match:
>       team: rocket
99d94
<     - team="rocket"
101c96
<   - receiver: team_ops_slack
---
>   - receiver: monitoring/alertmanager-config/team_ops_slack
266,267c261,262
< - name: root
< - name: harbor_implementation
---
> - name: monitoring/alertmanager-config/root
> - name: monitoring/alertmanager-config/harbor_implementation
288c283
< - name: falco_noise_slack
---
> - name: monitoring/alertmanager-config/falco_noise_slack
314c309
<       style: '{{ if eq .Status "firing" }}primary{{ else }}default{{ end }}'
---
>       style: default
327c322
<       url: '{{ template "__alert_silence_link" .}}'
---
>       url: '{{ template "__alert_silence_link" . }}'
329c324
< - name: team_atlas_slack
---
> - name: monitoring/alertmanager-config/team_atlas_slack
370c365
< - name: team_phoenix_slack
---
> - name: monitoring/alertmanager-config/team_bigmac_slack
377c372
<     channel: '#alert-phoenix-test'
---
>     channel: '#alert-bigmac-test'
409c404
<       url: '{{ template "__alert_silence_link" . }}'
---
>       url: '{{ template "__alert_silence_link" .}}'
411c406
< - name: team_hydra_slack
---
> - name: monitoring/alertmanager-config/team_clippy_slack
418c413
<     channel: '#alert-hydra-test'
---
>     channel: '#alert-clippy-test'
450c445
<       url: '{{ template "__alert_silence_link" . }}'
---
>       url: '{{ template "__alert_silence_link" .}}'
452c447
< - name: team_bigmac_slack
---
> - name: monitoring/alertmanager-config/team_hydra_slack
459c454
<     channel: '#alert-bigmac-test'
---
>     channel: '#alert-hydra-test'
491c486
<       url: '{{ template "__alert_silence_link" . }}'
---
>       url: '{{ template "__alert_silence_link" .}}'
493c488
< - name: team_clippy_slack
---
> - name: monitoring/alertmanager-config/team_phoenix_slack
500c495
<     channel: '#alert-clippy-test'
---
>     channel: '#alert-phoenix-test'
532c527
<       url: '{{ template "__alert_silence_link" . }}'
---
>       url: '{{ template "__alert_silence_link" .}}'
534c529
< - name: team_rocket_slack
---
> - name: monitoring/alertmanager-config/team_rocket_slack
573c568
<       url: '{{ template "__alert_silence_link" . }}'
---
>       url: '{{ template "__alert_silence_link" .}}'
575c570
< - name: team_shield_slack
---
> - name: monitoring/alertmanager-config/team_shield_slack
582c577
<     channel: '#alert-shield'
---
>     channel: '#alert-shield-test'
616c611
< - name: opsgenie_router
---
> - name: monitoring/alertmanager-config/opsgenie_router
631c626
< - name: team_ops_slack
---
> - name: monitoring/alertmanager-config/team_ops_slack
684c679
< - /etc/alertmanager/config/*.tmpl
---
> - /etc/alertmanager/templates/notification-template.tmpl

```

**Comment by QuentinBisson** [comment]
Sloth severity was removed because sloth alerts now support our old scheme https://github.com/giantswarm/sloth-rules/blob/f9139d349ddc59c9e0c50c6d5b584578da50a64f/helm/sloth-rules/files/kaas-phoenix-apiserver-availability.yaml#LL36C26-L36C26

**Comment by QuentinBisson** [comment]
@TheoBrigitte can you take a look?

**Comment by QuentinBisson** [comment]
There is a few things to fix in the notification template but the rest looks alright https://giantswarm.app.opsgenie.com/alert/detail/949b91a0-de51-468f-b272-fac3b87c39f1-1686606538844/details

**Comment by QuentinBisson** [comment]
I stand corrected the issue is related to mimir and not the migration as it's working on gremlin

File: helm/prometheus-meta-operator/values.yaml (hunk 1 above)
Comment (marieroque) [reply helm/prometheus-meta-operator/values.yaml:39]: Does it mean when a new team has routing alerts, we will have to release pmo ?

Comment (QuentinBisson) [reply helm/prometheus-meta-operator/values.yaml:39]: Yes it's how it's done today

File: helm/prometheus-meta-operator/files/notification-template.yaml [hunk 3]
```diff
@@ -56,10 +55,10 @@
...
     {{- end -}}
-    alertname%3D"{{ .CommonLabels.alertname }}"%7D
+        alertname%3D"{{ .CommonLabels.alertname }}"%7D
 {{- end }}
 
 # Link to related PMs
 {{ define "__alert_linked_postmortems" -}}
 https://github.com/giantswarm/giantswarm/issues?q=is%3Aissue+is%3Aopen+label%3Apostmortem+label%3Aalert%2F{{ .CommonLabels.alertname }}
-{{- end }}
+{{- end }}
```
Comment (QuentinBisson) [inline helm/prometheus-meta-operator/files/notification-template.yaml:64]:
```suggestion
{{- end }}

```

File: helm/prometheus-meta-operator/templates/alertmanager/alertmanager-global-secret.yaml [hunk 4]
```diff
@@ -7,5 +7,5 @@ metadata:
   namespace: {{ include "resource.default.namespace" . }}
 data:
   slackApiUrl: {{ .Values.alertmanager.slack.apiURL | b64enc | quote }}
-  opsGenieApiUrl: {{ printf "api.opsgenie.com" | b64enc | quote }}
+  opsGenieApiUrl: {{ printf "https://api.opsgenie.com/" | b64enc | quote }}
```
Comment (TheoBrigitte) [inline helm/prometheus-meta-operator/templates/alertmanager/alertmanager-global-secret.yaml:10]: I would rather have this in the values, to have things aligned.

**Review by TheoBrigitte — COMMENTED** [review]
Could we somehow keep the test file and test via helm ? I think this would help us get more confident that we are not breaking anything.

**Comment by QuentinBisson** [comment]
@TheoBrigitte I'm not even sure how you'd want me to test using helm

**Comment by QuentinBisson** [comment]
The issue is the format of the previous golden files and the new format is totally different (for the notification template we move to a configmap and for the alertmanager config we move to a CRD so we sould need to actually run the prometheus operator  to template a real alertmanager)  so there is no way to make sure te have the same values unless we diff the actual config in alertmanager as I did https://github.com/giantswarm/prometheus-meta-operator/pull/1316#issuecomment-1588148975

**Comment by TheoBrigitte** [comment]
Let's drop the golden/test files and move on with this PR as is.

**Pull request closed without merging**